lib/srv/db/mysql/autousers.go

```diff
@@ -0,0 +1,367 @@
+	//   row level security can be used in addition so each user can only read
+	//   their own attributes, if needed).
+	if isMariaDB(conn) {
+		return trace.NotImplemented("auto user provisioning is not supported for MariaDB yet")
+	}
+
+	// Create "teleport-auto-user".
+	_, err := conn.Execute(fmt.Sprintf("CREATE ROLE IF NOT EXISTS %q", teleportAutoUserRole))
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// There is no single command in MySQL to "CREATE OR REPLACE". Instead,
+	// have to DROP first before CREATE.
+	//
+	// To speed up the setup, the procedure "version" is stored as the
+	// procedure comment. So check if an update is necessary first by checking
+	// these comments.
+	//
+	// To force an update, drop one of the procedures or update the comment:
+	// ALTER PROCEDURE teleport_activate_user COMMENT 'need update'
+	if required, err := isProcedureUpdateRequired(conn, defaultSchema(sessionCtx), procedureVersion); err != nil {
+		return trace.Wrap(err)
+	} else if !required {
+		return nil
+	}
+
+	// If update is necessary, do a transaction.
+	e.Log.Debugf("Updating stored procedures for MySQL server %s.", sessionCtx.Database.GetName())
+	return trace.Wrap(doTransaction(conn, func() error {
+		for _, procedure := range allProcedures {
+			_, err := conn.Execute(fmt.Sprintf("DROP PROCEDURE IF EXISTS %s", procedure.name))
+			if err != nil {
+				return trace.Wrap(err)
+			}
+
+			_, err = conn.Execute(procedure.createCommand)
+			if err != nil {
+				return trace.Wrap(err)
+			}
+
+			_, err = conn.Execute(fmt.Sprintf("ALTER PROCEDURE %s COMMENT %q", procedure.name, procedureVersion))
+			if err != nil {
+				return trace.Wrap(err)
+			}
+		}
+		return nil
+	}))
+}
+
+// defaultSchema returns the default database to log into as the admin user.
+//
+// Use a default database/schema to make sure procedures are always created and
+// called from there (and possibly store other data there in the future).
+//
+// This also avoids "No database selected" errors if client doesn't provide
+// one.
+func defaultSchema(_ *common.Session) string {
+	// Use "mysql" as the default schema as both MySQL and Mariadb have it.
+	//
+	// TODO consider allowing user to specify the default database through database
+	// definition.
+	return "mysql"
+}
+
+func isMariaDB(conn *client.Conn) bool {
+	return strings.Contains(strings.ToLower(conn.GetServerVersion()), "mariadb")
+}
+
+// maxUsernameLength returns the username character limit.
+func maxUsernameLength(conn *client.Conn) int {
+	if isMariaDB(conn) {
+		return mariadbMaxUsernameLength
+	}
+	return mysqlMaxUsernameLength
+}
+
+func maybeHashUsername(teleportUser string, maxUsernameLength int) string {
+	if len(teleportUser) <= maxUsernameLength {
+		return teleportUser
+	}
+
+	// 64 bit hash collision rates:
+	// - 200 entries: 1 in 10^15
+	// -  2k entries: 1 in 10 trillion
+	// - 20k entries: 1 in 100 billion
+	hash := fnv.New64()
+	hash.Write([]byte(teleportUser))
```

Comment:
@jentfoo @reedloden 

Wonder about your thoughts about hashing teleport username. 

Namely, MySQL has hardcoded max user length limit (32 characters) but we want to support auto db user provisioning based on teleport username than can be longer than 32 character (for instance SSO login where username contains email address) 

The hashing approach  will allow us to support username > 32 but there is concern about hash collisions.

Reply:
It is actually relatively easy to cause collisions in `fnv` hash, but automatic username provisioning for DBs takes away the choice of the username. If the attacker has no control over their Teleport username, then a collision cannot be forced. 

Looking very briefly, at least some MySQL versions support "user attributes" in the form of custom JSON documents. Perhaps we could embed the complete expected username inside and check for collisions in this manner.

We could also have part of the username, as much as we can fit, embedded in the username: `t-{username_prefix}-{hash}`.

Reply:
Ah, actually I see we already have some code that deals with attributes, but MariaDB doesn't support those. A pity, but there might be some alternatives.

Reply:
I was thinking about this yesterday too. In general, even when two names collide, the user's roles get reset during activation. But it will become problematic when two names collide, and both have active connections. The 2nd user will automatically inherit 1st user's roles.

One way to solve this, like Krysztof suggested, we can throw an error when we detect collision. This can be easily done in the stored procedure using the original Teleport User from user attributes. 

> We could also have part of the username, as much as we can fit, embedded in the username: t-{username_prefix}-{hash}.

This can certainly reduce collision too. 

> https://github.com/gravitational/teleport-private/issues/998

Will follow up on that ticket. The solution will certainly have impact here

Reply:
Even with a `teleport-` prefix we can fit more than 128 bits of of data if we base 64 encode it.  That's not a ton of space to completely avoid conflicts, but it would be enough to support MD5.  Although still a weak algorithm it would be an improvement over FNV-1

A possibly better option is a shorter prefix of `tp-`.  That would give us enough space to encode a 160bit sha1 in base64 into the username

Reply:
@jentfoo that's awesome. I will update it to `tp-<base64-sha1-original-name>`. What do you think it's the best course of action when a collision does happen?

Reply:
@greedy52 I like your suggestion of:
> we can throw an error when we detect collision. This can be easily done in the stored procedure using the original Teleport User from user attributes. The only problem this way, is the 2nd user will never be able to login the database.

If we are using sha1 the chances of collision are so low that I think it's best to be defensive.